tests/basic/thread.rs

```diff
@@ -376,3 +377,119 @@ mod thread_local {
         )
     }
 }
+
+#[test]
+fn thread_park() {
+    check_dfs(
+        || {
+            let flag = Arc::new(AtomicBool::new(false));
+            let thd = {
+                let flag = Arc::clone(&flag);
+                thread::spawn(move || {
+                    thread::park();
+                    assert!(flag.load(Ordering::SeqCst));
+                })
+            };
+
+            flag.store(true, Ordering::SeqCst);
+            thd.thread().unpark();
+            thd.join().unwrap();
+        },
+        None,
+    )
+}
+
+#[test]
+#[should_panic(expected = "deadlock")]
+fn thread_park_deadlock() {
+    check_dfs(
+        || {
+            thread::park();
+        },
+        None,
+    )
+}
+
+// From the docs: "Because the token is initially absent, `unpark` followed by `park` will result in
+// the second call returning immediately"
+#[test]
+fn thread_unpark_park() {
+    check_dfs(
+        || {
+            thread::current().unpark();
+            thread::park();
+        },
+        None,
+    )
+}
+
+// Unparking a thread should not unconditionally unblock it (e.g., if it's blocked waiting on a lock
+// rather than parked)
+#[test]
+fn thread_unpark_unblock() {
+    check_dfs(
+        || {
+            let lock = Arc::new(Mutex::new(false));
+            let condvar = Arc::new(Condvar::new());
+
+            let reader = {
+                let lock = Arc::clone(&lock);
+                let condvar = Arc::clone(&condvar);
+                thread::spawn(move || {
+                    let mut guard = lock.lock().unwrap();
+                    while !*guard {
```

Comment:
Think a better version of this test would be to replace the `while` with an `if` and then add an `assert!(*guard)` afterwards.